src/include/duckdb/storage/compression/alp/algorithm/alp.hpp

```diff
@@ -0,0 +1,396 @@
+	 * Return TRUE if c1 is a better combination than c2
+	 * First criteria is number of times it appears as best combination
+	 * Second criteria is the estimated compression size
+	 * Third criteria is bigger exponent
+	 * Fourth criteria is bigger factor
+	 */
+	static bool CompareALPCombinations(const AlpCombination &c1, const AlpCombination &c2) {
+		return (c1.n_appearances > c2.n_appearances) ||
+		       (c1.n_appearances == c2.n_appearances &&
+		        (c1.estimated_compression_size < c2.estimated_compression_size)) ||
+		       ((c1.n_appearances == c2.n_appearances &&
+		         c1.estimated_compression_size == c2.estimated_compression_size) &&
+		        (c2.exponent < c1.exponent)) ||
+		       ((c1.n_appearances == c2.n_appearances &&
+		         c1.estimated_compression_size == c2.estimated_compression_size && c2.exponent == c1.exponent) &&
+		        (c2.factor < c1.factor));
+	}
+
+	/*
+	 * Dry compress a vector (ideally a sample) to estimate ALP compression size given a exponent and factor
+	 */
+	template <bool PENALIZE_EXCEPTIONS>
+	static uint64_t DryCompressToEstimateSize(const vector<T> &in, int8_t exp_idx, int8_t factor_idx) {
+		idx_t n_values = in.size();
+		idx_t exceptions_count = 0;
+		idx_t non_exceptions_count = 0;
+		uint32_t estimated_bits_per_value = 0;
+		uint64_t estimated_compression_size = 0;
+		int64_t max_encoded_value = NumericLimits<int64_t>::Minimum();
+		int64_t min_encoded_value = NumericLimits<int64_t>::Maximum();
+
+		for (const T &value : in) {
+			int64_t encoded_value = EncodeValue(value, exp_idx, factor_idx);
+			T decoded_value = DecodeValue(encoded_value, exp_idx, factor_idx);
+			if (decoded_value == value) {
+				non_exceptions_count++;
+				max_encoded_value = MaxValue(encoded_value, max_encoded_value);
+				min_encoded_value = MinValue(encoded_value, min_encoded_value);
+				continue;
+			}
+			exceptions_count++;
+		}
+
+		// We penalize combinations which yields to almost all exceptions
+		if (PENALIZE_EXCEPTIONS && non_exceptions_count < 2) {
+			return NumericLimits<uint64_t>::Maximum();
+		}
+
+		// Evaluate factor/exponent compression size (we optimize for FOR)
+		uint64_t delta = (static_cast<uint64_t>(max_encoded_value) - static_cast<uint64_t>(min_encoded_value));
+		estimated_bits_per_value = std::ceil(std::log2(delta + 1));
+		estimated_compression_size += n_values * estimated_bits_per_value;
+		estimated_compression_size +=
+		    exceptions_count * (EXACT_TYPE_BITSIZE + (AlpConstants::EXCEPTION_POSITION_SIZE * 8));
+		return estimated_compression_size;
+	}
+
+	/*
+	 * Find the best combinations of factor-exponent from each vector sampled from a rowgroup
+	 * This function is called once per segment
+	 * This operates over ALP first level samples
+	 */
+	static void FindTopKCombinations(const vector<vector<T>> &vectors_sampled, State &state) {
+		state.ResetCombinations();
+		// We use a 'pair' to hash it easily
+		map<pair<int8_t, int8_t>, uint64_t> best_k_combinations_hash;
```

Comment:
This sounds great.

We implemented a Hash function and an Equality function for `AlpEncodingIndices` to use it in an `ordered_map`; instead of the pair.